internal/traverse_map.go

```diff
@@ -0,0 +1,53 @@
+package internal
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
+
+// TraverseMap recursively walks the given map, calling set for each value. If the
+// value is a slice, set is called for each element of the slice. The keys of
+// nested maps are joined with the given delimiter.
+func TraverseMap(m map[string]any, delimiter string, set func(name, value string) error) error {
+	return traverseMap("", m, delimiter, set)
+}
+
+func traverseMap(key string, val any, delimiter string, set func(name, value string) error) error {
+	switch v := val.(type) {
+	case string:
+		return set(key, v)
+	case json.Number:
+		return set(key, v.String())
+	case uint64:
+		return set(key, strconv.FormatUint(v, 10))
+	case int:
+		return set(key, strconv.Itoa(v))
+	case int64:
+		return set(key, strconv.FormatInt(v, 10))
+	case float64:
+		return set(key, strconv.FormatFloat(v, 'g', -1, 64))
+	case bool:
+		return set(key, strconv.FormatBool(v))
+	case nil:
+		return set(key, "")
+	case []any:
+		for _, v := range v {
+			if err := traverseMap(key, v, delimiter, set); err != nil {
+				return err
+			}
+		}
+	case map[string]any:
```

Comment:
This will also require the change I made https://github.com/peterbourgon/ff/pull/108/commits/47743d74e9cc31bc9210a2b119db224ab20b83e0 to handle `map[any]any` which is what the YAML decoder uses as the default base type.

Reply:
Interesting, can you suggest a test case that would have caught this?

Reply:
A similar test as the one added in #108 
```yaml
foo:
  s: a string
  b: true
bar:
  nested:
    i: 123
    f: 1.23
baz:
  nested:
    x:
      - one
      - two
      - three
 ```